### PR TITLE
Adding AppProtocol to Service and Endpoints Ports

### PR DESCRIPTION
**Pull request by robscott**
**What type of PR is this?**
/kind api-change

**What this PR does / why we need it**:
This adds the `AppProtocol` field to Service and Endpoints resources, gated by a new `ServiceAppProtocol` feature gate.

**Does this PR introduce a user-facing change?**:
```release-note
AppProtocol is a new field on Service and Endpoints resources, enabled with the ServiceAppProtocol feature gate.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
- KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/20191227-app-protocol.md
- Enhancement Issue: https://github.com/kubernetes/enhancements/issues/1507

/sig network
/priority important-soon
/cc @liggitt 
/assign @thockin

### CODE REVIEW

File: pkg/apis/core/types.go [hunk 1]
```diff
@@ -3672,6 +3682,16 @@ type EndpointPort struct {
...
 	// The IP protocol for this port.
 	Protocol Protocol
+
+	// The application protocol for this port.
+	// This field follows standard Kubernetes label syntax.
+	// Un-prefixed names are reserved for IANA standard service names (as per
+	// RFC-6335 and http://www.iana.org/assignments/service-names).
+	// Non-standard protocols should use prefixed names such as
+	// mycompany.com/my-custom-protocol.
+	// Field can be enabled with ServiceAppProtocol Port.
```
Comment (thockin) [inline pkg/apis/core/types.go]: s/Port/Gate ?

File: pkg/apis/core/types.go [hunk 2]
```diff
@@ -3532,6 +3532,16 @@ type ServicePort struct {
 	// The IP protocol for this port.  Supports "TCP", "UDP", and "SCTP".
 	Protocol Protocol
 
+	// The application protocol for this port.
+	// This field follows standard Kubernetes label syntax.
+	// Un-prefixed names are reserved for IANA standard service names (as per
+	// RFC-6335 and http://www.iana.org/assignments/service-names).
+	// Non-standard protocols should use prefixed names such as
+	// mycompany.com/my-custom-protocol.
+	// Field can be enabled with ServiceAppProtocol Port.
```
Comment (thockin) [inline pkg/apis/core/types.go]: s/Port/Gate ?

**Review by thockin — COMMENTED** [review]
Thanks!

/lgtm
/approve

**Comment by robscott** [comment]
Thanks for the review and approval! Holding temporarily to fix my typo.

/hold

File: pkg/apis/core/types.go (hunk 1 above)
Comment (robscott) [reply pkg/apis/core/types.go]: Fixed, thanks!

File: pkg/apis/core/types.go (hunk 2 above)
Comment (robscott) [reply pkg/apis/core/types.go]: Fixed, thanks!

**Pull request merged**